### PR TITLE
Fix query escaping

### LINKED ISSUE #366: Query escaping
**Description**
`\GuzzleHttp\Psr7\Uri::withQueryValue(s)` does not handle `+` sign escaping;

**How to reproduce**
```php
$uri = \GuzzleHttp\Psr7\Uri::withQueryValue(
    new \GuzzleHttp\Psr7\Uri('test'),
    'date',
    '2021-01-11T17:50:00+03:00'
);
echo (string)$uri;
```
* Result: `test?date=2021-01-11T17:50:00+03:00`
* Expected result: `test?date=2021-01-11T17:50:00%2B03:00`

**Possible Solution**
Add `'+' => '%2B'` to `\GuzzleHttp\Psr7\Uri::$replaceQuery`?

**Additional context**
Also there is whole list of [reserved characters](https://en.wikipedia.org/wiki/Percent-encoding#Percent-encoding_reserved_characters), that must be percent-encoded.

### PR DESCRIPTION
**Pull request by red-led**
Fixes #366

### CODE REVIEW

File: phpstan-baseline.neon [hunk 1]
```diff
@@ -140,6 +140,11 @@ parameters:
 			count: 1
 			path: src/Uri.php
 
+		-
+			message: "#^Method GuzzleHttp\\\\Psr7\\\\Uri\\:\\:escapeQueryKey\\(\\) should return string but returns string\\|null\\.$#"
```
Comment (Nyholm) [inline phpstan-baseline.neon:144]: Why do you silence this?

Comment (GrahamCampbell) [reply phpstan-baseline.neon:144]: Because it's a false positive about the string replacement?

Comment (red-led) [reply phpstan-baseline.neon:144]: For same reason it's silenced for `filterPath` above. In theory `preg_replace_callback` returns `null` error. I can't imagine any error that can occur here. Regexp is valid, and input is guaranteed to be string by type hinting.

**Comment by red-led** [comment]
As a side note: there's a lot of messy guesswork involved in encodeing/decodeing queries. Consider dropping support of this behavior in future major versions, user should know if he is providing encoded query or not.

**Comment by Nyholm** [comment]
Graham, what do you think? Should we just rely on php' functions and ignore all the issues related to this. 

I would be positive to this.

**Comment by GrahamCampbell** [comment]
@Nyholm Yes, I agree. https://github.com/guzzle/psr7/issues/366#issuecomment-809988700 also raises a good point.

**Comment by GrahamCampbell** [comment]
Closing due to the discussion.

**Pull request closed without merging**